examples/albef/model.py

```diff
@@ -0,0 +1,385 @@
+            )
+            prediction_scores_m = self.answer_decoder_m(
+                input_ids=answers,
+                attention_mask=answers_atts,
+                encoder_hidden_states=encoder_hidden_states_m,
+                encoder_attention_mask=encoder_attention_mask,
+            )
+
+        prediction_scores = self.answer_decoder(
+            input_ids=answers,
+            attention_mask=answers_atts,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+        )
+
+        labels = answers.masked_fill(answers == 0, -100)
+        loss = self.loss(labels, prediction_scores, prediction_scores_m, alpha)
+        loss = ans_weights * loss
+        loss = loss.sum() / image.size(0)
+        return loss
+
+    def _test_forward(
+        self,
+        image: Tensor,
+        question: Tensor,
+        question_atts: Tensor,
+        answers: Tensor,
+        answer_atts: Tensor,
+        k: int = 128,
+    ) -> Tuple[Tensor, Tensor]:
+        encoder_outputs = self.model(image, question, question_atts)
+
+        num_ques = question.size(0)
+        start_ids = answers[0, 0].repeat(num_ques, 1)
+        atts = torch.ones(start_ids.shape)
+
+        prediction_scores = self.answer_decoder(
+            input_ids=start_ids,
+            attention_mask=atts,
+            encoder_hidden_states=encoder_outputs.multimodal_embeddings,
+            encoder_attention_mask=question_atts,
+        )
+
+        logits = prediction_scores[:, 0, :]  # get the cls token
+
+        answer_first_token = answers[:, 1]
+        prob_first_token = F.softmax(logits, dim=1).index_select(
+            dim=1, index=answer_first_token
+        )
+        topk_probs, topk_ids = prob_first_token.topk(k, dim=1)
+
+        input_ids = []
+        input_atts = []
+
+        for topk_id in topk_ids:
+            input_ids.append(answers.index_select(dim=0, index=topk_id))
+            input_atts.append(answer_atts.index_select(dim=0, index=topk_id))
+        input_ids = torch.cat(input_ids)
+        input_atts = torch.cat(input_atts)
+        targets_ids = input_ids.masked_fill(input_ids == 0, -100)
+
+        question_states = encoder_outputs.multimodal_embeddings.repeat_interleave(
+            k, dim=0
+        )
+        question_atts = question_atts.repeat_interleave(k, dim=0)
+
+        prediction_scores = self.answer_decoder(
+            input_ids=input_ids,
+            attention_mask=input_atts,
+            encoder_hidden_states=question_states,
+            encoder_attention_mask=question_atts,
+        )
+
+        answer_loss = self.loss(targets_ids, prediction_scores)
+        answer_loss = answer_loss.view(input_ids.size(0), -1)
+
+        # topk_prob: first token probability
+        topk_probs = topk_probs.view(-1, 1)
+        log_probs = torch.cat([topk_probs.log(), -answer_loss], dim=1)
+
+        # re-calculate log probabilities for the answer sequences using chain rule
```

Comment:
Not sure I understand this comment. Is this just about the product of conditional probabilities in the answer sequence?